ludwig/decoders/sequence_decoders.py

```diff
@@ -359,7 +441,14 @@ def decoder_teacher_forcing(
             [[0, 0], [0, 1], [0, 0]]
         )
 
-        return logits  # , outputs, final_state, generated_sequence_lengths
+        # extract structures needed for sampled_softmax
+        rnn_last_hidden = self.extract_decoder_end_state(final_state)
+
+        # EXPECTED SIZE OF RETURNED TENSORS
+        # logits: shape[batch_size, seq_size, num_classes]
+        # rnn_last_hidden: shape[batch_size, state_size] or
+        #     with Attention shape[batch_size, state_size * num_layers]
```

Comment:
why is it `shape[batch_size, state_size * num_layers]` ? We shoudl probably just use the one from the last layer

Reply:
With the implementation of Ludwig's custom BasicDecoder class, the `rnn_last_hidden` is now shaped as [batch_size, seq_size, state_size].  This is the tensor that is input into the final Dense layer of the output feature.  Actually, the next change will rename `run_last_hidden` to `projection_input` to be more explicit on its use.